libraries/Adapters/Microsoft.Bot.Builder.Adapters.Slack/SlackHelper.cs

```diff
@@ -0,0 +1,356 @@
+                }
+
+                if (attachments.Count > 0)
+                {
+                    message.Attachments = attachments;
+                }
+            }
+
+            message.Channel = activity.Conversation.Id;
+
+            if (!string.IsNullOrWhiteSpace(activity.Conversation.Properties["thread_ts"]?.ToString()))
+            {
+                message.ThreadTs = activity.Conversation.Properties["thread_ts"].ToString();
+            }
+
+            // if channelData is specified, overwrite any fields in message object
+            if (activity.ChannelData != null)
+            {
+                message = activity.GetChannelData<NewSlackMessage>();
+            }
+
+            // should this message be sent as an ephemeral message
+            if (!string.IsNullOrWhiteSpace(message.Ephemeral))
+            {
+                message.User = activity.Recipient.Id;
+            }
+
+            if (message.IconUrl != null || !string.IsNullOrWhiteSpace(message.Icons?.status_emoji) || !string.IsNullOrWhiteSpace(message.Username))
+            {
+                message.AsUser = false;
+            }
+
+            return message;
+        }
+
+        /// <summary>
+        /// Writes the HttpResponse.
+        /// </summary>
+        /// <param name="response">The httpResponse.</param>
+        /// <param name="code">The status code to be written.</param>
+        /// <param name="text">The text to be written.</param>
+        /// <param name="encoding">The encoding for the text.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public static async Task WriteAsync(HttpResponse response, HttpStatusCode code, string text, Encoding encoding, CancellationToken cancellationToken = default)
+        {
+            if (response == null)
+            {
+                throw new ArgumentNullException(nameof(response));
+            }
+
+            if (text == null)
+            {
+                throw new ArgumentNullException(nameof(text));
+            }
+
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            response.ContentType = "text/plain";
+            response.StatusCode = (int)code;
+
+            var data = encoding.GetBytes(text);
+
+            await response.Body.WriteAsync(data, 0, data.Length, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Creates an activity based on the slack event payload.
+        /// </summary>
+        /// <param name="slack">The payload of the slack event.</param>
+        /// <returns>An activity containing the event data.</returns>
+        public static Activity PayloadToActivity(SlackPayload slack)
+        {
+            if (slack == null)
+            {
+                return null;
```

Comment:
Should this ever be null?  If not why not those ArgumentNullException?

Reply:
Same as above. Please, let us know if you consider that the exception is the right approach here and we'll change it.

Reply:
As above. Spoken we Chris / Gabo. This should be throwing the ArgumentNullException and not returning null. Please amend. Thanks.

Reply:
Done. Thanks!